kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/SniAwareClusterEndpointConfigProvider.java

```diff
@@ -0,0 +1,102 @@
+
+public class SniAwareClusterEndpointConfigProvider implements ClusterEndpointConfigProvider {
+
+    private static final EndpointMatchResult BOOTSTRAP_MATCHED = new EndpointMatchResult(true, null);
+    private static final EndpointMatchResult NO_MATCH = new EndpointMatchResult(false, null);
+    private static final Pattern NODE_ID_TOKEN = Pattern.compile("\\$\\(nodeId\\)");
+    public static final String LITERAL_NODE_ID = "$(nodeId)";
+    private final HostPort bootstrapAddress;
+    private final String brokerAddressPattern;
+    private final Pattern brokerAddressPatternRegExp;
+
+    public SniAwareClusterEndpointConfigProvider(SniAwareClusterEndpointProviderConfig config) {
+        this.bootstrapAddress = config.bootstrapAddress;
+        this.brokerAddressPattern = config.brokerAddressPattern;
+        this.brokerAddressPatternRegExp = config.brokerAddressPatternRegExp;
+    }
+
+    @Override
+    public HostPort getClusterBootstrapAddress() {
+        return this.bootstrapAddress;
+    }
+
+    @Override
+    public HostPort getBrokerAddress(int nodeId) {
+        if (nodeId < 0) {
+            // nodeIds of < 0 have special meaning to kafka.
+            throw new IllegalArgumentException("nodeId cannot be less than zero");
+        }
+        return new HostPort(brokerAddressPattern.replace(LITERAL_NODE_ID, Integer.toString(nodeId)), bootstrapAddress.port());
+    }
+
+    @Override
+    public EndpointMatchResult hasMatchingEndpoint(String sniHostname, int port) {
+        if (sniHostname == null || bootstrapAddress.port() != port) {
+            return NO_MATCH;
+        }
+        else if (bootstrapAddress.host().equals(sniHostname) || bootstrapAddress.host().toLowerCase(Locale.ROOT).equals(sniHostname.toLowerCase(Locale.ROOT))) {
+            return BOOTSTRAP_MATCHED;
+        }
+        var matcher = brokerAddressPatternRegExp.matcher(sniHostname);
+        if (matcher.find()) {
+            return new EndpointMatchResult(true, Integer.parseInt(matcher.group(1)));
+        }
+        return NO_MATCH;
+    }
+
+    @Override
+    public boolean requiresPortExclusivity() {
+        return false;
+    }
+
+    @Override
+    public boolean requiresTls() {
+        return true;
+    }
+
+    public static class SniAwareClusterEndpointProviderConfig extends BaseConfig {
+        private final Pattern brokerAddressPatternRegExp;
+
+        private final HostPort bootstrapAddress;
+        private final String brokerAddressPattern;
+
+        public SniAwareClusterEndpointProviderConfig(HostPort bootstrapAddress, String brokerAddressPattern) {
+            if (bootstrapAddress == null) {
+                throw new IllegalArgumentException("bootstrapAddress cannot be null");
+            }
+            if (brokerAddressPattern == null) {
+                throw new IllegalArgumentException("brokerAddressPattern cannot be null");
+            }
+
+            var matcher = NODE_ID_TOKEN.matcher(brokerAddressPattern);
```

Comment:
suggest replacing the `replaceFirst` check with:

```
long matches = matcher.results().count();
            if (matches != 1) {
                throw new IllegalArgumentException("brokerAddressPattern must contain exactly one nodeId replacement pattern " + LITERAL_NODE_ID + ". Found " + matches);
            }
```

Reply:
I used `replaceFirst` as a cheap way of disallowing things like `"toomanyrecursive$(nodeId$(nodeId))"`.  I don't think your replacement does that

I am foreseeing that for the next phase #271 we'll need the ability to interpolate other tokens such as `$(port)` into broker address patterns.  I think that'll be the point to introduce proper parsing.